chain/consensus/mir/manager.go

```diff
@@ -0,0 +1,355 @@
+	// TODO: Configure repo path to persist Mir WAL
+	path := fmt.Sprintf("lotus-wal%s", strings.Replace(mirID, "/", "-", -1))
+	wal, err := NewWAL(mirID, path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create WAL: %w", err)
+	}
+
+	netTransport, err := mirlibp2p.NewTransport(h, t.NodeID(mirID), logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create transport: %w", err)
+	}
+	if err := netTransport.Start(); err != nil {
+		return nil, fmt.Errorf("failed to start transport: %w", err)
+	}
+	netTransport.Connect(ctx, initialMembership)
+
+	cryptoManager, err := NewCryptoManager(addr, api)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create crypto manager: %w", err)
+	}
+
+	// Instantiate an interceptor.
+	var interceptor *eventlog.Recorder
+
+	m := Manager{
+		Addr:                addr,
+		NetName:             netName,
+		Pool:                fifo.New(),
+		MirID:               mirID,
+		interceptor:         interceptor,
+		WAL:                 wal,
+		CryptoManager:       cryptoManager,
+		Net:                 netTransport,
+		InitialValidatorSet: initialValidatorSet,
+		ToMir:               make(chan chan []*mirproto.Request),
+	}
+
+	m.StateManager = NewStateManager(initialMembership, &m)
+
+	mpool := pool.NewModule(
+		m.ToMir,
+		&pool.ModuleConfig{
+			Self:   "mempool",
+			Hasher: "hasher",
+		},
+		&pool.ModuleParams{
+			MaxTransactionsInBatch: 10,
+		},
+	)
+
+	smrSystem, err := smr.New(
+		t.NodeID(mirID),
+		h,
+		initialMembership,
+		m.CryptoManager,
+		m.StateManager,
+		logger,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not create SMR system: %w", err)
+	}
+	smrSystem = smrSystem.WithModule("mempool", mpool)
+
+	if err := smrSystem.Start(ctx); err != nil {
+		return nil, fmt.Errorf("could not start SMR system: %w", err)
+	}
+
+	cfg := mir.DefaultNodeConfig().WithLogger(logger)
+
+	interceptorOutput := os.Getenv(InterceptorOutputEnv)
+	if interceptorOutput != "" {
+		// TODO: Persist in repo path?
+		m.interceptor, err = eventlog.NewRecorder(
+			t.NodeID(mirID),
+			interceptorOutput,
+			logging.Decorate(logger, "Interceptor: "),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create interceptor: %w", err)
+		}
+		m.MirNode, err = mir.NewNode(t.NodeID(mirID), cfg, smrSystem.Modules(), wal, m.interceptor)
+	} else {
+		m.MirNode, err = mir.NewNode(t.NodeID(mirID), cfg, smrSystem.Modules(), wal, nil)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Mir node: %w", err)
+	}
+
+	return &m, nil
+}
+
+// Start starts the manager.
+func (m *Manager) Start(ctx context.Context) chan error {
+	log.Infof("Mir manager %s starting", m.MirID)
+
+	errChan := make(chan error, 1)
+
+	go func() {
+		// Run Mir node until it stops.
+		if err := m.MirNode.Run(ctx); err != nil && !errors.Is(err, mir.ErrStopped) {
+			log.Infof("Mir manager %s: Mir node stopped with error: %v", m.MirID, err)
+			errChan <- err
+		}
+
+		// Perform cleanup of Node's modules.
+		m.Stop()
```

Comment:
Good question. The convention I usually use is that the "cleaning order" is inverse of how you created the objects (following an "ownership" approach). 

In our case the manager owns the node, thus, the node is closed first and then the manager closes. If you think this makes no sense we can consider changing it.

Reply:
Ah yes I see your point. That makes most sense when the manager is not being stopped from the outside, i.e., when some high-level code wants to stop the it will need to tell the Mir node (which lives inside the manager) to stop, and then wait for the manager to stop automatically. In which case we probably shouldn't export the manager's `Stop` function.

I completely agree with the stack-based cleaning order, but for me that would translate to something like
```
func manager.Stop() {
    mirNode.Stop()
    clean up manager
    return
}
```

Reply:
I guess the question basically was whether to call `m.Stop()` from inside the manager when the included Mir node "spontaneously" stops. But I don't have strong feelings about this, if that's convenient for integration, just do whatever you think appropriate and feel free to resolve this conversation without any further comment.

Reply:
@dnkolegov, that is a great point. Now that I look at the code again, I would also use this approach: 
```
func manager.Stop() {
    mirNode.Stop()
    clean up manager
    return
}
```
@dnkolegov, any strong reason not to move `mirNode.Stop` from `manager.Start` to `manager.Stop`?

Reply:
Are we talking just about adding 
`m.MirNode.Stop()` to manager's Stop?
Because nobody will call the manager's Stop beside defer.

```
func (m *Manager) Stop() {
	log.With("miner", m.MirID).Infof("Mir manager shutting down")
	defer log.With("miner", m.MirID).Info("Mir manager stopped")

	m.MirNode.Stop()

	if m.interceptor != nil {
		if err := m.interceptor.Stop(); err != nil {
			log.Errorf("Could not close interceptor: %s", err)
		}
		log.Info("Interceptor closed")
	}

	m.Net.Stop()
	log.Info("Network transport stopped")
}
```

Reply:
That was the idea (@matejpavlovic, correct me if I'm wrong), but no strong feelings for now, we can keep it as-is if it works.

Reply:
Yes, that was the idea. But as I said, no big deal.

Reply:
fixed